File: ios/Sources/SummaryScreen.swift
import SwiftUI
import BorgarlandCore

/// The POC ends here: every field that would be posted, displayed exactly.
/// The send step exists (unlike the Android POC, which had no networking
/// dependency at all) but posts only to our relay, which is in dry run and
/// forwards nothing. Decision 0002 put that decision on the server so it is
/// a deploy away, not an App Store release.
struct SummaryScreen: View {
    @ObservedObject var model: ReportModel
    let payload: Payload

    /// The title, pinned above the scroll view rather than carried inside it.
    ///
    /// `safeAreaInset` does two things at once and both are the fix: it puts an
    /// opaque strip between the status bar and the content, and it insets the
    /// scroll view's own safe area so the content stops at the strip instead of
    /// passing under it. In the field-test screenshot the clock sat on top of a
    /// category name and both were unreadable (#78).
    private func header(_ title: String) -> some View {
        Text(title)
            .font(.title2)
            .fontWeight(.semibold)
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.horizontal, 16)
            .padding(.top, 16)
            .padding(.bottom, 8)
            .background(Color(.systemBackground))
    }

    var body: some View {
        let state = model.state
        ScrollView {
            VStack(alignment: .leading, spacing: 0) {
                Text("Ekkert fer til borgarinnar. Þetta app sendir aðeins á okkar relay (ákvörðun 0002) og relay-ið er í þurrkeyrslu og framsendir ekkert. Hér fyrir neðan er nákvæmlega það sem færi yfir línuna.")
                    .font(.footnote)
                    .padding(12)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .background(Color(.secondarySystemBackground), in: RoundedRectangle(cornerRadius: 8))
                    .padding(.top, 12)

                if state.outOfBounds {
                    Text("Viðvörun: hnitin falla utan þess svæðis sem kort borgarinnar sýnir.")
                        .foregroundStyle(.red)
                        .padding(.top, 8)
                }

                FieldRow(label: "category", value: payload.categorySlug)
                FieldRow(label: "latitude", value: payload.latitudeText)
                FieldRow(label: "longitude", value: payload.longitudeText)
                FieldRow(label: "description", value: payload.description)

                Text("photo")
                    .font(.subheadline)
                    .fontWeight(.semibold)
                    .padding(.top, 12)

                // By offset, not by name: every photo the app produces is
                // called mynd.jpg, and the contract allows the photo part to
                // repeat, so names are not identities here.
                ForEach(Array(payload.photos.enumerated()), id: \.offset) { _, file in
                    HStack(spacing: 12) {
                        if let image = PhotoBytes.image(from: file.bytes) {
                            image
                                .resizable()
                                .scaledToFill()
                                .frame(width: 64, height: 64)
                                .clipShape(RoundedRectangle(cornerRadius: 8))
                        }
                        VStack(alignment: .leading, spacing: 2) {
                            Text(file.name)
                            Text("\(file.mime), \(file.sizeBytes) bytes")
                                .font(.caption)
                                .foregroundStyle(.secondary)
                        }
                    }
                    .padding(.vertical, 6)
                }

                Text("Staðsetning úr: \(state.locationSource ?? "")")
                    .font(.caption)
                    .padding(.top, 12)

                Button {
                    model.sendToRelay()
                } label: {
                    HStack(spacing: 8) {
                        // Something moving, so the wait is legible as a wait
                        // rather than as a dead screen (#73).
                        if state.sending {
                            ProgressView()
                        }
                        Text(state.sending ? "Sendi..." : "Senda á relay")
                    }
                    .frame(maxWidth: .infinity)
                }
                .buttonStyle(.borderedProminent)
                .padding(.top, 16)
                .disabled(state.sending)

                // The way out. Before #73 the send control was dead for
                // eighty-four seconds and the only live thing beside it was
                // Byrja aftur, which abandoned the report instead of stopping
                // the request. This stops the request; the report waits.
                if state.sending {
                    Button {
                        model.cancelSend()
                    } label: {
                        Text("Hætta við")
                            .frame(maxWidth: .infinity)
                    }
                    .buttonStyle(.bordered)
                    .padding(.top, 8)
                }

                // Where the report goes, in words rather than a hostname. A URL tells
                // a reader nothing they can act on, and the loopback one told them
                // something false (#29).
                Text("Sendist á þjónustu Borgarlands, ekki beint til borgarinnar. Þjónustan er í þurrkeyrslu og framsendir ekkert.")
                    .font(.caption)
                    .padding(.top, 8)

                if let note = state.deliveryNote {
                    VStack(alignment: .leading, spacing: 8) {
                        Text(note)
                            .font(.footnote)
                        // The one deliberate way a report leaves the phone
                        // without reaching the relay. Spelled out, because
                        // leaving a screen must not be how a report is thrown
                        // away.
                        if state.currentReportIsQueued {
                            Button(role: .destructive) {
                                model.discardCurrentReport()
                            } label: {
                                Text("Eyða ábendingunni")
                            }
                            .buttonStyle(.bordered)
                        }
                    }
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(12)
                    .background(Color(.secondarySystemBackground), in: RoundedRectangle(cornerRadius: 8))
                    .padding(.top, 12)
                }

                if let sent = state.sendOutcome {
                    VStack(alignment: .leading, spacing: 8) {
                        // The sentence first, and in the person's language. A
                        // tester read the raw refusal off this screen and asked
                        // what it meant (#77).
                        if let outcome = sent.outcome {
                            Text(outcome.says)
                                .font(.subheadline)
                                .fontWeight(.semibold)
                            if let advice = outcome.advice {
                                Text(advice)
                                    .font(.footnote)
                            }
                        }

                        // Still reachable, and still exactly what the relay
                        // said: it is the fastest way to see what happened, and
                        // losing it would trade one defect for another. Behind
                        // a control the person opens on purpose, so the default
                        // screen is a sentence rather than a diagnostic dump.
                        DisclosureGroup("Tæknilegt svar") {
                            Text(sent.raw)
                                .font(.system(.footnote, design: .monospaced))
                                .textSelection(.enabled)
                                .frame(maxWidth: .infinity, alignment: .leading)
                                .padding(.top, 8)
                        }
                        .font(.footnote)
                    }
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(12)
                    .background(Color(.secondarySystemBackground), in: RoundedRectangle(cornerRadius: 8))
                    .padding(.top, 12)
                }

                Button {
                    model.startOver()
                } label: {
                    Text("Byrja aftur")
                        .frame(maxWidth: .infinity)
                }
                .buttonStyle(.bordered)
                .padding(.top, 16)
            }
            .padding(16)
        }
        // The same exposure as the details screen, never photographed doing it
        // but identical in shape (#78).
        .safeAreaInset(edge: .top, spacing: 0) { header("Það sem yrði sent") }
    }
}

/// The Kotlin's FieldRow: a divider, the wire field name in small secondary
/// type, and the value in monospace. The wire names are shown raw to the
/// user on purpose — this screen documents the contract.
private struct FieldRow: View {
    let label: String
    let value: String

    var body: some View {
        VStack(alignment: .leading, spacing: 2) {
            Divider()
                .padding(.vertical, 8)
            Text(label)
                .font(.caption)
                .foregroundStyle(.secondary)
            Text(value)
                .font(.system(.body, design: .monospaced))
        }
    }
}

File: android/app/src/main/kotlin/is/borgarland/ui/SummaryScreen.kt
package `is`.borgarland.ui

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.unit.dp
import `is`.borgarland.Payload
import `is`.borgarland.PocUiState

/**
 * Every field that would be posted, displayed exactly, and then the send.
 *
 * This comment used to say "nothing is sent, ever ... no INTERNET permission,
 * no HTTP client, no networking dependency". All three were true when the
 * screen was the end of the POC and none of them survived the relay: the
 * manifest declares INTERNET, RelayClient posts over HttpURLConnection, and
 * PocViewModel.sendToRelay is wired to the button below.
 *
 * What is still true is the reason the screen exists: a person sees exactly
 * what will be sent before it is sent. The relay is in dry run and forwards
 * nothing to the city (worker/src/config.ts), which is a property of the
 * server and not of this app.
 */
@Composable
fun SummaryScreen(
    state: PocUiState,
    payload: Payload,
    onStartOver: () -> Unit,
    onSend: () -> Unit,
) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .verticalScroll(rememberScrollState())
            .padding(16.dp),
    ) {
        Text("Það sem yrði sent", style = MaterialTheme.typography.headlineMedium)

        Card(
            modifier = Modifier.fillMaxWidth().padding(top = 12.dp),
            colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.secondaryContainer),
        ) {
            Text(
                "Ekkert fer til borgarinnar. Þetta app sendir aðeins á okkar relay (ákvörðun 0002) og relay-ið er í þurrkeyrslu og framsendir ekkert. Hér fyrir neðan er nákvæmlega það sem færi yfir línuna.",
                style = MaterialTheme.typography.bodyMedium,
                modifier = Modifier.padding(12.dp),
            )
        }

        if (state.outOfBounds) {
            Text(
                "Viðvörun: hnitin falla utan þess svæðis sem kort borgarinnar sýnir.",
                color = MaterialTheme.colorScheme.error,
                style = MaterialTheme.typography.bodyMedium,
                modifier = Modifier.padding(top = 8.dp),
            )
        }

        // First, because it is first in data/relay-request.json and because it
        // is the one field on this screen that identifies THIS report rather
        // than describing it. The screen is headed "what would be sent" and had
        // been omitting a part that is sent (#88 put it on the wire; nothing
        // put it on the screen).
        //
        // Readable matters as well as present: the id is generated when this
        // screen is reached and kept until the walk starts over, so it is the
        // one value a person can read out BEFORE sending — which is what lets
        // an operator authorise one specific report rather than one device.
        FieldRow("reportId", payload.reportId ?: "—")
        FieldRow("category", payload.categorySlug)
        FieldRow("latitude", payload.latitudeText)
        FieldRow("longitude", payload.longitudeText)
        FieldRow("description", payload.description)

        Text(
            "photo",
            style = MaterialTheme.typography.titleSmall,
            modifier = Modifier.padding(top = 12.dp),
        )
        payload.photos.forEach { file ->
            Row(
                modifier = Modifier.fillMaxWidth().padding(vertical = 6.dp),
            ) {
                val bitmap = remember(file.bytes, file.rotationDegrees) { decodePhoto(file.bytes, file.rotationDegrees) }
                Image(
                    bitmap = bitmap.asImageBitmap(),
                    contentDescription = file.name,
                    modifier = Modifier.size(64.dp).clip(RoundedCornerShape(8.dp)),
                    contentScale = ContentScale.Crop,
                )
                Spacer(Modifier.width(12.dp))
                Column {
                    Text(file.name, style = MaterialTheme.typography.bodyLarge)
                    Text(
                        "${file.mime}, ${file.sizeBytes} bytes",
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
            }
        }

        Text(
            "Staðsetning úr: ${state.locationSource.orEmpty()}",
            style = MaterialTheme.typography.bodySmall,
            modifier = Modifier.padding(top = 12.dp),
        )

        // Disabled once the relay has taken it. The button used to return to
        // its unsent state after a 201, fully live, as though nothing had been
        // sent — and a tester pressed it again and filed the same ábending
        // twice (#85). The relay now refuses to store the repeat (#88); this is
        // so nobody is invited to make one.
        val alreadySent = state.sendOutcome?.ok == true
        // And disabled once the answer cannot change. A refusal a person
        // retries is a refusal that does not read as final: a tester pressed
        // send three times against the same jurisdiction 400 because nothing
        // on the screen said the coordinate was the problem and the button was
        // the only live control (#148). Which answers are terminal is
        // data/relay-outcomes.json's call, not this screen's.
        val refusedForGood = state.sendOutcome?.outcome?.retryable == false
        Button(
            onClick = onSend,
            enabled = !state.sending && !alreadySent && !refusedForGood,
            modifier = Modifier.fillMaxWidth().padding(top = 16.dp),
        ) {
            Text(
                when {
                    state.sending -> "Sendi..."
                    alreadySent -> "Sent"
                    else -> "Senda á relay"
                },
            )
        }

        Text(
            // The app cannot reach the city. It posts to our relay, which is
            // in dry run and forwards nothing. Decision 0002 put that decision
            // on the server so it is a deploy away, not an App Store release.
            // Where the report goes, in words rather than a hostname. A URL tells
            // a reader nothing they can act on, and the loopback one told them
            // something false (#29).
            "Sendist á þjónustu Borgarlands, ekki beint til borgarinnar. " +
                "Þjónustan er í þurrkeyrslu og framsendir ekkert.",
            style = MaterialTheme.typography.bodySmall,
            modifier = Modifier.padding(top = 8.dp),
        )

        state.sendOutcome?.let { sent ->
            var rawShown by remember(sent) { mutableStateOf(false) }
            Card(
                modifier = Modifier.fillMaxWidth().padding(top = 12.dp),
                colors = CardDefaults.cardColors(
                    containerColor = MaterialTheme.colorScheme.secondaryContainer),
            ) {
                Column(modifier = Modifier.padding(12.dp)) {
                    // The sentence first, and in the person's language. A
                    // tester read the raw refusal off this screen and asked
                    // what it meant (#77).
                    sent.outcome?.let { outcome ->
                        Text(outcome.says, style = MaterialTheme.typography.titleSmall)
                        outcome.advice?.let { advice ->
                            Text(
                                advice,
                                style = MaterialTheme.typography.bodySmall,
                                modifier = Modifier.padding(top = 4.dp),
                            )
                        }
                        // Where the person actually is, when the relay said.
                        // The refusal used to name what we would not do and
                        // never where the coordinate fell (#148). Null when the
                        // relay sent no such field, and then there is no line at
                        // all rather than one with a hole in it.
                        outcome.detail?.let { detail ->
                            Text(
                                detail,
                                style = MaterialTheme.typography.bodySmall,
                                modifier = Modifier.padding(top = 4.dp),
                            )
                        }
                    }

                    // Still reachable, and still exactly what the relay said:
                    // it is the fastest way to see what happened, and losing it
                    // would trade one defect for another. Behind a control the
                    // person opens on purpose, so the default screen is a
                    // sentence rather than a diagnostic dump.
                    TextButton(
                        onClick = { rawShown = !rawShown },
                        modifier = Modifier.padding(top = 4.dp),
                    ) {
                        Text(if (rawShown) "Fela tæknilegt svar" else "Tæknilegt svar")
                    }
                    if (rawShown) {
                        Text(
                            sent.raw,
                            fontFamily = FontFamily.Monospace,
                            style = MaterialTheme.typography.bodySmall,
                        )
                    }
                }
            }
        }

        Button(
            onClick = onStartOver,
            modifier = Modifier.fillMaxWidth().padding(top = 16.dp),
        ) { Text("Byrja aftur") }
    }
}

@Composable
private fun FieldRow(label: String, value: String) {
    HorizontalDivider(modifier = Modifier.padding(vertical = 8.dp))
    Text(label, style = MaterialTheme.typography.labelMedium, color = MaterialTheme.colorScheme.onSurfaceVariant)
    Text(
        value,
        style = MaterialTheme.typography.bodyLarge.copy(fontFamily = FontFamily.Monospace),
    )
}

private fun decodePhoto(bytes: ByteArray, rotationDegrees: Int): Bitmap {
    val decoded = BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
        ?: return Bitmap.createBitmap(1, 1, Bitmap.Config.ARGB_8888)
    if (rotationDegrees % 360 == 0) return decoded
    return Bitmap.createBitmap(
        decoded, 0, 0, decoded.width, decoded.height,
        Matrix().apply { postRotate(rotationDegrees.toFloat()) }, true,
    )
}
